File: sdks/go/pomi/classifier.go
package pomi

import (
	"math"
	"regexp"
	"sort"
	"strconv"
	"strings"

	xagentauth "github.com/dyshay/agentauth/sdks/go"
)

// ClassifierOptions configures the model classifier.
type ClassifierOptions struct {
	ConfidenceThreshold float64
}

// ModelClassifier performs Bayesian inference to identify the model family
// based on canary responses.
type ModelClassifier struct {
	modelFamilies       []string
	confidenceThreshold float64
	extractor           *CanaryExtractor
}

// NewModelClassifier creates a new ModelClassifier.
func NewModelClassifier(modelFamilies []string, options *ClassifierOptions) *ModelClassifier {
	threshold := 0.5
	if options != nil && options.ConfidenceThreshold > 0 {
		threshold = options.ConfidenceThreshold
	}
	return &ModelClassifier{
		modelFamilies:       modelFamilies,
		confidenceThreshold: threshold,
		extractor:           NewCanaryExtractor(),
	}
}

// Classify performs Bayesian classification on the canary responses.
func (mc *ModelClassifier) Classify(
	canaries []xagentauth.Canary,
	canaryResponses map[string]string,
) xagentauth.ModelIdentification {
	if canaryResponses == nil || len(canaries) == 0 {
		return xagentauth.ModelIdentification{
			Family: "unknown", Confidence: 0, Evidence: nil, Alternatives: nil,
		}
	}

	evidence := mc.extractor.Extract(canaries, canaryResponses)
	if len(evidence) == 0 {
		return xagentauth.ModelIdentification{
			Family: "unknown", Confidence: 0, Evidence: nil, Alternatives: nil,
		}
	}

	// Initialize uniform prior
	posteriors := make(map[string]float64, len(mc.modelFamilies))
	for _, family := range mc.modelFamilies {
		posteriors[family] = 1.0 / float64(len(mc.modelFamilies))
	}

	// Bayesian update for each canary with a response
	for _, canary := range canaries {
		response, ok := canaryResponses[canary.ID]
		if !ok {
			continue
		}

		for _, family := range mc.modelFamilies {
			prior := posteriors[family]
			likelihood := mc.computeLikelihood(canary, response, family)
			posteriors[family] = prior * likelihood
		}

		// Normalize after each update to prevent underflow
		mc.normalize(posteriors)
	}

	// Find the best hypothesis
	bestFamily := "unknown"
	bestConfidence := 0.0

	for family, posterior := range posteriors {
		if posterior > bestConfidence {
			bestConfidence = posterior
			bestFamily = family
		}
	}

	// Build alternatives (sorted descending, excluding best)
	var alternatives []xagentauth.ModelAlternative
	for family, posterior := range posteriors {
		if family != bestFamily {
			alternatives = append(alternatives, xagentauth.ModelAlternative{
				Family:     family,
				Confidence: math.Round(posterior*1000) / 1000,
			})
		}
	}
	sort.Slice(alternatives, func(i, j int) bool {
		return alternatives[i].Confidence > alternatives[j].Confidence
	})

	// Apply confidence threshold
	if bestConfidence < mc.confidenceThreshold {
		return xagentauth.ModelIdentification{
			Family:     "unknown",
			Confidence: math.Round(bestConfidence*1000) / 1000,
			Evidence:   evidence,
			Alternatives: append([]xagentauth.ModelAlternative{
				{Family: bestFamily, Confidence: math.Round(bestConfidence*1000) / 1000},
			}, alternatives...),
		}
	}

	return xagentauth.ModelIdentification{
		Family:       bestFamily,
		Confidence:   math.Round(bestConfidence*1000) / 1000,
		Evidence:     evidence,
		Alternatives: alternatives,
	}
}

func (mc *ModelClassifier) computeLikelihood(canary xagentauth.Canary, response string, family string) float64 {
	weight := canary.ConfidenceWeight

	switch canary.Analysis.Type {
	case "exact_match":
		expected, ok := canary.Analysis.Expected[family]
		if !ok {
			return 0.5
		}
		isMatch := strings.EqualFold(strings.TrimSpace(response), strings.TrimSpace(expected))
		if isMatch {
			return 0.5 + 0.5*weight
		}
		return 0.5 - 0.4*weight

	case "pattern":
		pattern, ok := canary.Analysis.Patterns[family]
		if !ok {
			return 0.5
		}
		re, err := regexp.Compile("(?i)" + pattern)
		if err != nil {
			return 0.5
		}
		if re.MatchString(response) {
			return 0.5 + 0.45*weight
		}
		return 0.5 - 0.35*weight

	case "statistical":
		dist, ok := canary.Analysis.Distributions[family]
		if !ok {
			return 0.5
		}
		re := regexp.MustCompile(`-?\d+\.?\d*`)
		numMatch := re.FindString(response)
		if numMatch == "" {
			return 0.5
		}
		value, err := strconv.ParseFloat(numMatch, 64)
		if err != nil {
			return 0.5
		}
		pdf := gaussianPdf(value, dist.Mean, dist.StdDev)
		maxPdf := gaussianPdf(dist.Mean, dist.Mean, dist.StdDev)
		normalizedPdf := pdf / maxPdf
		return 0.1 + 0.8*normalizedPdf*weight
	}
	return 0.5
}

func gaussianPdf(x, mean, stddev float64) float64 {
	z := (x - mean) / stddev
	return math.Exp(-0.5*z*z) / (stddev * math.Sqrt(2*math.Pi))
}

func (mc *ModelClassifier) normalize(posteriors map[string]float64) {
	sum := 0.0
	for _, v := range posteriors {
		sum += v
	}
	if sum == 0 {
		for k := range posteriors {
			posteriors[k] = 1.0 / float64(len(posteriors))
		}
		return
	}
	for k, v := range posteriors {
		posteriors[k] = v / sum
	}
}

File: sdks/rust/src/pomi/classifier.rs
use std::collections::HashMap;

use regex::Regex;

use crate::types::{Canary, CanaryAnalysis, ModelAlternative, ModelIdentification};

use super::extractor::CanaryExtractor;

/// Bayesian model classifier using canary response analysis.
pub struct ModelClassifier {
    model_families: Vec<String>,
    confidence_threshold: f64,
    extractor: CanaryExtractor,
}

impl ModelClassifier {
    pub fn new(model_families: Vec<String>, confidence_threshold: f64) -> Self {
        Self {
            model_families,
            confidence_threshold,
            extractor: CanaryExtractor::new(),
        }
    }

    /// Classify the model family based on canary responses using Bayesian inference.
    pub fn classify(
        &self,
        canaries: &[Canary],
        canary_responses: &Option<HashMap<String, String>>,
    ) -> ModelIdentification {
        if canary_responses.is_none() || canaries.is_empty() {
            return ModelIdentification {
                family: "unknown".into(),
                confidence: 0.0,
                evidence: vec![],
                alternatives: vec![],
            };
        }

        let responses = canary_responses.as_ref().unwrap();
        let evidence = self.extractor.extract(canaries, canary_responses);

        if evidence.is_empty() {
            return ModelIdentification {
                family: "unknown".into(),
                confidence: 0.0,
                evidence: vec![],
                alternatives: vec![],
            };
        }

        // Initialize uniform prior
        let n = self.model_families.len() as f64;
        let mut posteriors: HashMap<String, f64> = self
            .model_families
            .iter()
            .map(|f| (f.clone(), 1.0 / n))
            .collect();

        // Bayesian update for each canary with a response
        for canary in canaries {
            let response = match responses.get(&canary.id) {
                Some(r) => r,
                None => continue,
            };

            for family in &self.model_families {
                let prior = *posteriors.get(family).unwrap_or(&0.0);
                let likelihood = self.compute_likelihood(canary, response, family);
                posteriors.insert(family.clone(), prior * likelihood);
            }

            // Normalize after each update to prevent underflow
            self.normalize(&mut posteriors);
        }

        // Find the best hypothesis
        let mut best_family = "unknown".to_string();
        let mut best_confidence = 0.0_f64;

        for (family, posterior) in &posteriors {
            if *posterior > best_confidence {
                best_confidence = *posterior;
                best_family = family.clone();
            }
        }

        // Build alternatives
        let mut alternatives: Vec<ModelAlternative> = posteriors
            .iter()
            .filter(|(f, _)| **f != best_family)
            .map(|(f, p)| ModelAlternative {
                family: f.clone(),
                confidence: (p * 1000.0).round() / 1000.0,
            })
            .collect();
        alternatives.sort_by(|a, b| b.confidence.partial_cmp(&a.confidence).unwrap());

        // Apply confidence threshold
        if best_confidence < self.confidence_threshold {
            return ModelIdentification {
                family: "unknown".into(),
                confidence: (best_confidence * 1000.0).round() / 1000.0,
                evidence,
                alternatives: {
                    let mut alts = vec![ModelAlternative {
                        family: best_family,
                        confidence: (best_confidence * 1000.0).round() / 1000.0,
                    }];
                    alts.extend(alternatives);
                    alts
                },
            };
        }

        ModelIdentification {
            family: best_family,
            confidence: (best_confidence * 1000.0).round() / 1000.0,
            evidence,
            alternatives,
        }
    }

    fn compute_likelihood(&self, canary: &Canary, response: &str, family: &str) -> f64 {
        let weight = canary.confidence_weight;

        match &canary.analysis {
            CanaryAnalysis::ExactMatch { expected } => {
                let exp = match expected.get(family) {
                    Some(e) => e,
                    None => return 0.5,
                };
                let is_match = response.trim().to_lowercase() == exp.trim().to_lowercase();
                if is_match {
                    0.5 + 0.5 * weight
                } else {
                    0.5 - 0.4 * weight
                }
            }
            CanaryAnalysis::Pattern { patterns } => {
                let pattern = match patterns.get(family) {
                    Some(p) => p,
                    None => return 0.5,
                };
                let is_match = match Regex::new(&format!("(?i){}", pattern)) {
                    Ok(re) => re.is_match(response),
                    Err(_) => return 0.5,
                };
                if is_match {
                    0.5 + 0.45 * weight
                } else {
                    0.5 - 0.35 * weight
                }
            }
            CanaryAnalysis::Statistical { distributions } => {
                let dist = match distributions.get(family) {
                    Some(d) => d,
                    None => return 0.5,
                };
                let re = Regex::new(r"-?\d+\.?\d*").unwrap();
                let value = match re
                    .find(response)
                    .and_then(|m| m.as_str().parse::<f64>().ok())
                {
                    Some(v) => v,
                    None => return 0.5,
                };
                let pdf = self.gaussian_pdf(value, dist.mean, dist.stddev);
                let max_pdf = self.gaussian_pdf(dist.mean, dist.mean, dist.stddev);
                let normalized = if max_pdf > 0.0 { pdf / max_pdf } else { 0.0 };
                0.1 + 0.8 * normalized * weight
            }
        }
    }

    fn gaussian_pdf(&self, x: f64, mean: f64, stddev: f64) -> f64 {
        let z = (x - mean) / stddev;
        (-0.5 * z * z).exp() / (stddev * (2.0 * std::f64::consts::PI).sqrt())
    }

    fn normalize(&self, posteriors: &mut HashMap<String, f64>) {
        let sum: f64 = posteriors.values().sum();
        if sum == 0.0 {
            let n = posteriors.len() as f64;
            for v in posteriors.values_mut() {
                *v = 1.0 / n;
            }
            return;
        }
        for v in posteriors.values_mut() {
            *v /= sum;
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::types::InjectionMethod;

    fn default_families() -> Vec<String> {
        vec![
            "gpt-4-class".into(),
            "claude-3-class".into(),
            "gemini-class".into(),
            "llama-class".into(),
            "mistral-class".into(),
        ]
    }

    fn make_canary() -> Canary {
        Canary {
            id: "test".into(),
            prompt: "test".into(),
            injection_method: InjectionMethod::Inline,
            analysis: CanaryAnalysis::ExactMatch {
                expected: [
                    ("gpt-4-class".to_string(), "alpha".to_string()),
                    ("claude-3-class".to_string(), "beta".to_string()),
                ]
                .into_iter()
                .collect(),
            },
            confidence_weight: 0.5,
        }
    }

    #[test]
    fn test_classify_known() {
        let classifier = ModelClassifier::new(default_families(), 0.3);
        let canary = make_canary();
        let mut responses = HashMap::new();
        responses.insert("test".to_string(), "alpha".to_string());

        let result = classifier.classify(&[canary], &Some(responses));
        // Should identify gpt-4-class since "alpha" matches it
        assert!(result.confidence > 0.0);
        // The family should be either gpt-4-class or something else depending on Bayesian outcome
    }

    #[test]
    fn test_classify_unknown() {
        let classifier = ModelClassifier::new(default_families(), 0.9);
        let canary = make_canary();
        let mut responses = HashMap::new();
        responses.insert("test".to_string(), "gamma".to_string());

        let result = classifier.classify(&[canary], &Some(responses));
        // With high threshold, "gamma" won't match, so likely "unknown"
        // The confidence might be below threshold
        assert!(!result.evidence.is_empty() || result.family == "unknown");
    }

    #[test]
    fn test_classify_no_responses() {
        let classifier = ModelClassifier::new(default_families(), 0.5);
        let canary = make_canary();

        let result = classifier.classify(&[canary], &None);
        assert_eq!(result.family, "unknown");
        assert_eq!(result.confidence, 0.0);
    }
}
